source/opt/simplification_pass.cpp

```diff
@@ -0,0 +1,114 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "simplification_pass.h"
+
+#include <set>
+#include <unordered_set>
+#include <vector>
+
+#include "fold.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status SimplificationPass::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+  bool modified = false;
+
+  for (ir::Function& function : *get_module()) {
+    modified |= SimplifyFunction(&function);
+  }
+  return (modified ? Status::SuccessWithChange : Status::SuccessWithoutChange);
+}
+
+bool SimplificationPass::SimplifyFunction(ir::Function* function) {
+  bool modified = false;
+  // Phase 1: Traverse all instruction in dominance order.
+  // The second phase will only be on the instructions whose input have changed
+  // after being processed during phase 1.  Since OpPhi instructions that the
+  // only instructions whose inputs do not necessarily dominate the use, we keep
+  // track of the OpPhi instructions already seen, and add them to the work list
+  // for phase 2 when needed.
+  std::vector<ir::Instruction*> work_list;
+  std::unordered_set<ir::Instruction*> process_phis;
+  std::set<ir::Instruction*> inst_to_kill;
+
+  opt::DominatorAnalysis* dom =
+      context()->GetDominatorAnalysis(function, *context()->cfg());
+
+  dom->Visit([&modified, &process_phis, &work_list,
+              this](DominatorTreeNode* dom_node) {
+    for (ir::Instruction* inst = &*dom_node->bb_->begin(); inst;) {
+      if (inst->opcode() == SpvOpPhi) {
+        process_phis.insert(inst);
+      }
+
+      if (FoldInstruction(inst)) {
+        modified = true;
+        context()->AnalyzeUses(inst);
+        get_def_use_mgr()->ForEachUser(
+            inst, [&work_list, &process_phis](ir::Instruction* use) {
+              if (process_phis.count(use)) {
+                work_list.push_back(use);
+              }
+            });
+        if (inst->opcode() == SpvOpCopyObject) {
+          context()->ReplaceAllUsesWith(inst->result_id(),
+                                        inst->GetSingleWordInOperand(0));
+          inst = context()->KillInst(inst);
+        } else {
+          inst = inst->NextNode();
+        }
+      } else {
+        inst = inst->NextNode();
+      }
+    }
+    return true;
+  });
+
+  // Phase 2: process the instructions in the work list until all of the work is
+  //          done.  This time we add all users to the work list because phase 1
+  //          has already finished.
+
+  // Using a set to make sure entries are unique, and the traversal order is
+  // deterministic.
+  while (!work_list.empty()) {
+    ir::Instruction* inst = work_list.back();
+    work_list.pop_back();
+    if (FoldInstruction(inst)) {
+      modified = true;
+      context()->AnalyzeUses(inst);
+      get_def_use_mgr()->ForEachUser(inst, [&work_list](ir::Instruction* use) {
+        work_list.push_back(use);
```

Comment:
Prevent multiple copies of the same user from being enqueued on the worklist? Might have to change to a breadth first traversal.